sortstructs/priority-sorted-struct.go

```diff
@@ -0,0 +1,148 @@
+	elements []map[int]K
+	// numberOfPriorities is the number of priorities in the priorityList + 1
+	numberOfPriorities int
+	// currentElementNumber is the current element number
+	currentElementNumber int
+}
+
+// PriorityIndex is the priority and index used to locate items
+type PriorityIndex struct {
+	Priority int
+	Index    int
+}
+
+func NewPrioritySortedStruct[P comparable, K any](priorityList [][]P, priorityMatchFunction func(el K) P) *PrioritySortedStruct[P, K] {
+	p := PrioritySortedStruct[P, K]{
+		priorityList:     priorityList,
+		getPriorityValue: priorityMatchFunction,
+	}
+	p.Init()
+	return &p
+}
+
+func (p *PrioritySortedStruct[P, K]) Init() {
+	p.numberOfPriorities = len(p.priorityList)
+	p.elements = make([]map[int]K, p.numberOfPriorities+1)
+	if p.priorityList != nil {
+		for priorityIndex := range p.priorityList {
+			p.elements[priorityIndex] = make(map[int]K)
+		}
+		p.elements[p.numberOfPriorities] = make(map[int]K)
+	} else {
+		// there is only 1 map of elements
+		p.elements[0] = make(map[int]K)
+	}
+	p.priorityMap = make(map[P]int)
+	for index, pl := range p.priorityList {
+		for _, v := range pl {
+			p.priorityMap[v] = index
+		}
+	}
+	p.count = 0
+	p.currentElementNumber = 0
+}
+
+// Get returns the element at the index, and if it exists
+func (p *PrioritySortedStruct[P, K]) Get(pi PriorityIndex) (K, bool) {
+	v, ok := p.elements[pi.Priority][pi.Index]
+	return v, ok
+}
+
+// Process will call the procesFunc over all the elements by priority
+func (p *PrioritySortedStruct[P, K]) Process(processFunc func(el K, pi PriorityIndex)) {
+	for i := 0; i <= p.numberOfPriorities; i++ {
+		m := p.elements[i]
+		for index, v := range m {
+			processFunc(v, PriorityIndex{Priority: i, Index: index})
+		}
+	}
+}
+
+// GetPriorityList returns an ordered list of the elements by priority
+func (p *PrioritySortedStruct[P, K]) GetPriorityList() []K {
+	elements := make([]K, p.count)
+	currentIndex := 0
+	for priority := 0; priority <= p.numberOfPriorities; priority++ {
+		mapOfElements := p.elements[priority]
+		for _, el := range mapOfElements {
+			elements[currentIndex] = el
+			currentIndex++
+		}
+	}
+	return elements
+}
+
+// Add will add the element to the Priority Collection, returns the priority, and element number.
+func (p *PrioritySortedStruct[P, K]) Add(element K) PriorityIndex {
+	priority := p.getPriorityOfElement(element)
+	if priority == defaultPrioirty {
+		// add to the last index of the watches
+		priority = p.numberOfPriorities
+	}
+	p.elements[priority][p.currentElementNumber] = element
+	pi := PriorityIndex{Priority: priority, Index: p.currentElementNumber}
+	p.count++
+	p.currentElementNumber++
+	return pi
+}
+
+// Delete will delete the element, return if it deleted
+func (p *PrioritySortedStruct[P, K]) Delete(pi PriorityIndex) bool {
+	if p.count == 0 {
+		return false
+	}
+	if _, ok := p.Get(pi); ok {
+		delete(p.elements[pi.Priority], pi.Index)
+		p.count--
```

Comment:
This could cause 2 elements to be added that have the same index. I will change the name of the variable to `currentElementIndex`. This way it will maintain its meaning.